<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>XenSCE: exploration of a class for Xenium demonstration data • XenSCE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="XenSCE: exploration of a class for Xenium demonstration data">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">XenSCE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/XenSCE.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/XenSCE/">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="XenSCE_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>XenSCE: exploration of a class for Xenium demonstration data</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">September 07, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/XenSCE/blob/HEAD/vignettes/XenSCE.Rmd"><code>vignettes/XenSCE.Rmd</code></a></small>
      <div class="hidden name"><code>XenSCE.Rmd</code></div>

    </div>

    
    
<style type="text/css">
blockquote {
    padding: 10px 20px;
    margin: 0 0 20px;
    font-size: 14px;
    border-left: 5px solid #eee;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h3>
<p>Numerous groups are working on methodology for analyzing new data resources emerging from spatial transcriptomics experiments. The “computing technology stack” necessary to work with such data is complex. The following <a href="https://spatialdata.scverse.org/en/latest/tutorials/notebooks/notebooks/examples/models1.html" class="external-link">comment</a> (Aug 16 2024) from the scverse group is characteristic:</p>
<blockquote>
<p>The SpatialData Zarr format, which is described in our design doc, is an extension of the OME-NGFF specification, which makes use of the OME-Zarr, the AnnData Zarr and the Parquet file formats. We need to use these combination of technologies because currently OME-NGFF does not provide all the fundamentals required for storing spatial omics dataset; nevertheless, we try to stay as close as OME-NGFF as possible, and we are contributing to ultimately make spatial omics support available in pure OME-NGFF.</p>
</blockquote>
<p>In Bioconductor 3.19, we might be committed to all these technologies (which include but do not mention HDF5) plus R, reticulate, basilisk, and sf to have a widely functional solution. Attaching the <a href="https://bioconductor.org/packages/Voyager" class="external-link">Voyager</a> package leads to loading over 100 associated packages. The XenSCE package is intended to explore the balance between functionality and high dependency load specifically for the analysis of outputs from 10x Xenium experiments. The XenSCE class extends <a href="https://bioconductor.org/packages/SpatialExperiment" class="external-link">SpatialExperiment</a> to optionally include references to parquet files that manage voluminous geometry data; geometry can also be handled as DataFrame for sufficiently low volume experiments.</p>
<p>As of 0.0.7 of XenSCE, a new class, XenSPEP, is provided to address parquet representation of cell, nucleus and transcript coordinates.</p>
</div>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>This package is based on publicly available datasets.</p>
<ul>
<li><p><strong>Pancreas sample</strong>. Obtainable via <code>data(panc_sub)</code> this is derived from <code>SpatialFeatureExperiment::readXenium(SFEData::XeniumOutput("v2"))</code> but the assay has been replaced with an equivalent dgCMatrix instance, and the geometry data are provided as thin DataFrame instances (thin meaning each coordinate has a separate row), whereas the SpatialFeatureExperiment uses simple features to organize geometry information on cells, nuclei, and transcripts.</p></li>
<li><p><strong>Human brain</strong> <a href="https://www.10xgenomics.com/datasets/ffpe-human-brain-cancer-data-with-human-immuno-oncology-profiling-panel-and-custom-add-on-1-standard" class="external-link">FFPE Glioblastoma Data</a>. We have taken several components of the distribution and placed them in an NSF Open Storage Network bucket. The <code>cache_assets</code> function in this package will download and cache about 4GB of compressed sparse matrix and parquet data files; it returns the paths to these resources. The <code>build_demo</code> function creates an instance of <code>XenSCE</code>.</p></li>
<li><p><strong>Human lung (SFE-based)</strong> <a href="https://www.10xgenomics.com/datasets/preview-data-ffpe-human-lung-cancer-with-xenium-multimodal-cell-segmentation-1-standard" class="external-link">FFPE Lung cancer dataset</a>. We used scripts distributed by scverse to create a zarr-based representation, and used <code>readXenium</code> on the 10x distribution. Use <code>example(sfe2xsce)</code> to retrieve and cache and transform the SpatialFeatureExperiment to a XenSCE instance.</p></li>
<li><p><strong>Human dermal melanoma</strong> <a href="https://www.10xgenomics.com/datasets/xenium-prime-ffpe-human-skin" class="external-link">FFPE Human pan tissue dataset</a>. For this dataset, we retrieved the <code>outs.zip</code> file from this site and ran <code><a href="../reference/ingest_xen.html">XenSCE::ingest_xen</a></code> on the contents, producing a XenSPEP instance which was then saved in an HDF5-backed representation. This, along with the parquet files for cell, nucleus, and transcript coordinates, are zipped together and placed in an Open Storage Network bucket for retrieval via <code>cache_xen_sk</code>.</p></li>
<li><p><strong>Human prostate adenocarcinoma</strong> <a href="https://www.10xgenomics.com/datasets/xenium-prime-ffpe-human-prostate" class="external-link">FFPE Human pan tissue dataset</a>. For this dataset, we retrieved the <code>outs.zip</code> file from this site and ran <code><a href="../reference/ingest_xen.html">XenSCE::ingest_xen</a></code> on the contents, producing a XenSPEP instance which was then saved in an HDF5-backed representation. This, along with the parquet files for cell, nucleus, and transcript coordinates, are zipped together and placed in an Open Storage Network bucket for retrieval via <code>cache_xen_prost</code>.</p></li>
<li><p><strong>Human Lung (SPEP example)</strong> Again working with <a href="https://www.10xgenomics.com/datasets/preview-data-ffpe-human-lung-cancer-with-xenium-multimodal-cell-segmentation-1-standard" class="external-link">FFPE Lung cancer preview</a>, but working with <code>ingest_xen</code> and <code>zipXenSPEP</code> to produce a portable archive. Use <code>example(cache_xen_luad)</code> to obtain this instance of XenSPEP.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="a-quick-look">A quick look<a class="anchor" aria-label="anchor" href="#a-quick-look"></a>
</h3>
<div class="section level4">
<h4 id="small-in-memory-example">Small in-memory example<a class="anchor" aria-label="anchor" href="#small-in-memory-example"></a>
</h4>
<p>Boundary coordinates are in memory, in S4Vector::DataFrame instances.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/XenSCE">XenSCE</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">panc_sub</span><span class="op">)</span></span>
<span><span class="va">panc_sub</span></span></code></pre></div>
<pre><code><span><span class="co">## class: XenSCE </span></span>
<span><span class="co">## dim: 514 6272 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(514): ENSG00000121270 ENSG00000130234 ...</span></span>
<span><span class="co">##   UnassignedCodeword_0497 UnassignedCodeword_0499</span></span>
<span><span class="co">## rowData names(1): symbol</span></span>
<span><span class="co">## colnames(6272): abclkehb-1 abcnopgp-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">## colData names(12): cell_id x_centroid ... nucleus_area sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">## imgData names(0):</span></span>
<span><span class="co">## Geometry elements:</span></span>
<span><span class="co">##               nrow ncol</span></span>
<span><span class="co">## transcripts 367354   11</span></span>
<span><span class="co">## cellbounds  156512    4</span></span>
<span><span class="co">## nucbounds   154156    4</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">SpatialExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">panc_sub</span><span class="op">)</span>, </span>
<span>   pch<span class="op">=</span><span class="st">"."</span> , cex<span class="op">=</span><span class="fl">1.5</span>, xlab<span class="op">=</span><span class="st">"x"</span>, ylab<span class="op">=</span><span class="st">"y"</span>, main<span class="op">=</span><span class="st">"cell centroids"</span><span class="op">)</span></span></code></pre></div>
<p><img src="XenSCE_files/figure-html/dolit-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="larger-hybrid-example">Larger hybrid example<a class="anchor" aria-label="anchor" href="#larger-hybrid-example"></a>
</h4>
<p>In this example, transcript counts are in memory in a sparse matrix, but geometry information is handled in parquet. The <code>view_seg_g2</code> function allows selection of two gene symbols, and cells are colored according to single or double occupancy.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">luad</span> <span class="op">=</span> <span class="fu"><a href="../reference/cache_xen_luad.html">cache_xen_luad</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pa</span> <span class="op">=</span> <span class="fu"><a href="../reference/cache_xen_luad.html">cache_xen_luad</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">luad</span> <span class="op">=</span> <span class="fu"><a href="../reference/restoreZipXenSPEP.html">restoreZipXenSPEP</a></span><span class="op">(</span><span class="va">pa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">luad</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/make.names.html" class="external-link">make.names</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">luad</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span>, unique<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">=</span> <span class="fu"><a href="../reference/view_seg_g2.html">view_seg_g2</a></span><span class="op">(</span><span class="va">luad</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5800</span>, <span class="fl">6300</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1300</span>, <span class="fl">1800</span><span class="op">)</span>, lwd<span class="op">=</span><span class="fl">.5</span>, gene1<span class="op">=</span><span class="st">"CD4"</span>, gene2<span class="op">=</span><span class="st">"EPCAM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">5800</span>,<span class="fl">1390</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"purple"</span>, <span class="st">"cyan"</span>, <span class="st">"pink"</span><span class="op">)</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"EPCAM"</span>, <span class="st">"both"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="XenSCE_files/figure-html/dolung-1.png" width="672"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span><span class="op">$</span><span class="va">ncells</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2074</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="interactivity">Interactivity<a class="anchor" aria-label="anchor" href="#interactivity"></a>
</h4>
<p>In <code>inst/app4</code>, code is provided to work with the primary dermal melanoma dataset. A map of the cell coordinates can drive focused exploration. The region of interest is shown by a whitened rectangle in the upper left corner.</p>
<p><img src="fblmap.png"></p>
<p>Cells are colored by quintile of size. Points where FBL transcripts are found are plotted as dots.</p>
<p><img src="FBLsketch.png"></p>
<p>More work is needed to identify useful exploratory visualizations.</p>
</div>
<div class="section level4">
<h4 id="large-disk-based-example">Large disk-based example<a class="anchor" aria-label="anchor" href="#large-disk-based-example"></a>
</h4>
<div class="section level5">
<h5 id="hdf5parquet-backed-examples">HDF5+parquet-backed examples<a class="anchor" aria-label="anchor" href="#hdf5parquet-backed-examples"></a>
</h5>
<p>We want to be able to accommodate very large numbers of cells and transcripts without heavy infrastructure commitments. Examples with the XenSPEP class illustrate the current approach.</p>
<p>We’ll use the prostate adenocarcinoma 5K dataset to demonstrate. A 900MB zip file will be cached.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prost</span> <span class="op">=</span> <span class="fu"><a href="../reference/cache_xen_prost.html">cache_xen_prost</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><code>prost</code> is the path to a zip file in a BiocFileCache instance.</p>
<p>Create a folder to work in, and unzip.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">xpw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"prost_work"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">prost</span>, exdir<span class="op">=</span><span class="va">xpw</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">xpw</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "cell_boundaries.parquet"    "nucleus_boundaries.parquet"</span></span>
<span><span class="co">## [3] "README.txt"                 "transcripts.parquet"       </span></span>
<span><span class="co">## [5] "xen_prost"</span></span></code></pre>
<p>Restore the SpatialExperiment component.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prostx</span> <span class="op">=</span> <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/saveHDF5SummarizedExperiment.html" class="external-link">loadHDF5SummarizedExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">xpw</span>, <span class="st">"xen_prost"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prostx</span></span></code></pre></div>
<pre><code><span><span class="co">## XenSPEP instance.  SCEcomponent:</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 5006 193000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(3): ID Symbol Type</span></span>
<span><span class="co">## colnames(193000): aaaagkdm-1 aaaamcnn-1 ... oioglipl-1 oiohaagd-1</span></span>
<span><span class="co">## colData names(15): cell_id x_centroid ... segmentation_method sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: Gene Expression</span></span>
<span><span class="co">## altExpNames(5): Deprecated Codeword Genomic Control Negative Control</span></span>
<span><span class="co">##   Codeword Negative Control Probe Unassigned Codeword</span></span>
<span><span class="co">## use spatialCoords() for cell centroids.</span></span>
<span><span class="co">## Geometry element paths (not loaded):</span></span>
<span><span class="co">##  ./cell_boundaries.parquet</span></span>
<span><span class="co">##  ./nucleus_boundaries.parquet</span></span>
<span><span class="co">##  ./transcripts.parquet</span></span></code></pre>
</div>
<div class="section level5">
<h5 id="approaching-obsolescence-gbm">Approaching obsolescence: GBM<a class="anchor" aria-label="anchor" href="#approaching-obsolescence-gbm"></a>
</h5>
<p>If you are interested in working with the 10X GBM data, use the following:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/XenSCE">XenSCE</a></span><span class="op">)</span></span>
<span><span class="va">myxen</span> <span class="op">=</span> <span class="fu"><a href="../reference/retrieve_demo.html">retrieve_demo</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># steps noted above were done</span></span>
<span><span class="va">myxen</span></span></code></pre></div>
<p>To sample the cell centroid coordinates and plot, use:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sinds</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">myxen</span><span class="op">)</span><span class="op">)</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">myxen</span><span class="op">$</span><span class="va">x_centroid</span><span class="op">[</span><span class="va">sinds</span><span class="op">]</span>, <span class="op">-</span><span class="va">myxen</span><span class="op">$</span><span class="va">y_centroid</span><span class="op">[</span><span class="va">sinds</span><span class="op">]</span>,</span>
<span>  xlab<span class="op">=</span><span class="st">"x"</span>, ylab<span class="op">=</span><span class="st">"y"</span>, main<span class="op">=</span><span class="st">"Sample of 5000 centroids"</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span></span></code></pre></div>
<p>Compare the basic layout to the ‘summary’ visualization shipped with the data.</p>
<div class="figure">
<img src="qcimg.png" alt=""><p class="caption">from site</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="clipping-and-visualizing">Clipping and visualizing<a class="anchor" aria-label="anchor" href="#clipping-and-visualizing"></a>
</h3>
<p><code>clip_rect</code> is an endomorphism that restricts the cells in an XenSCE instance to those with centroids in given boundaries in x and y. By default, cell centroid and transcript locations are added. AT PRESENT (0.0.15) THIS ONLY WORKS FOR IN-MEMORY/NON-PARQUET-BASED ExAMPLES.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBoundaries.html">plotCellBoundaries</a></span><span class="op">(</span><span class="fu"><a href="../reference/clip_rect.html">clip_rect</a></span><span class="op">(</span><span class="va">panc_sub</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>,<span class="fl">850</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>,<span class="fl">750</span><span class="op">)</span><span class="op">)</span>, add_tx<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="XenSCE_files/figure-html/dosimpa-1.png" width="576"></p>
<p>With transcript locations:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBoundaries.html">plotCellBoundaries</a></span><span class="op">(</span><span class="fu"><a href="../reference/clip_rect.html">clip_rect</a></span><span class="op">(</span><span class="va">panc_sub</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>,<span class="fl">850</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>,<span class="fl">750</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="XenSCE_files/figure-html/dosimp-1.png" width="576"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
